emcache/cluster.py

```diff
@@ -71,42 +82,75 @@ def __init__(
         ssl: bool,
         ssl_verify: bool,
         ssl_extra_ca: Optional[str],
+        autodiscovery: bool,
+        autodiscovery_poll_interval: float,
+        autodiscovery_timeout: float,
+        loop: asyncio.AbstractEventLoop,
     ) -> None:
 
         if not memcached_hosts_address:
             raise ValueError("Nodes can't be an empty list, at least one node has to be provided")
 
         self._cluster_events = cluster_events
         self._purge_unhealthy_nodes = purge_unhealthy_nodes
+        self._autodiscovery = autodiscovery
+        self._autodiscovery_poll_interval = autodiscovery_poll_interval
+        self._autodiscovery_timeout = autodiscovery_timeout
+        self._autodiscover_config_version = -1
+        self._first_autodiscovery_done = loop.create_future()
+        self._loop = loop
 
         # Create nodes and configure them to be used by the Rendezvous
         # hashing. By default all are placed under the healthy list, if later
         # on they report that are unhealhty and `purge_unhealthy_nodes` has been
         # configured they will be removed.
-        self._unhealthy_nodes = []
-        self._healthy_nodes = [
-            Node(
-                memcached_host_address,
-                max_connections,
-                min_connections,
-                purge_unused_connections_after,
-                connection_timeout,
-                self._on_node_healthy_status_change_cb,
-                ssl,
-                ssl_verify,
-                ssl_extra_ca,
-            )
-            for memcached_host_address in memcached_hosts_address
+        self._new_node_options = [
+            max_connections,
+            min_connections,
+            purge_unused_connections_after,
+            connection_timeout,
+            self._on_node_healthy_status_change_cb,
+            ssl,
+            ssl_verify,
+            ssl_extra_ca,
         ]
-        self._build_rdz_nodes()
         self._cluster_managment = _ClusterManagment(self)
         self._events = asyncio.Queue(maxsize=MAX_EVENTS)
         self._events_dispatcher_task = asyncio.get_running_loop().create_task(self._events_dispatcher())
         self._closed = False
+
+        if self._autodiscovery:
+            self._unhealthy_nodes = []
+            self._healthy_nodes = []
+            self._discovery_nodes = [
+                Node(
+                    memcached_host_address,
+                    1,
+                    1,
+                    None,
+                    connection_timeout,
+                    None,
+                    ssl,
+                    ssl_verify,
+                    ssl_extra_ca,
+                )
+                for memcached_host_address in memcached_hosts_address
+            ]
```

Comment:
Originally I used copy of `self._healthy_nodes` and created `self._original_nodes` and used the later one for discovery, but kept the seed node in `self._healthy_nodes`. The biggest reason is my fear of getting into situation where somehow all nodes were marked as unhealthy and I don't have anything to look up.

Also the configuration hostname `mycluster.fnjyzo.cfg.use1.cache.amazonaws.com` looks very different than the original hostnames for example `0001.mycluster.fnjyzo.use1.cache.amazonaws.com`, `0002.mycluster.fnjyzo.use1.cache.amazonaws.com`. It also behaves differently. The configuration hostname has very short TTL and rotates over different nodes.

When I would add it to `self._healthy_nodes` it would screw up the hashing and make one node be used more than the others. The issue also got bigger when I increased minimum connections, as then I noticed I was making many connections to that seed node, so it was clear this needs to be treated differently, so this and how it affected hashing was the main reason for the change.

Reply:
Good point. thanks for the explanation!